src/combobox/ComboBox.ts

```diff
@@ -0,0 +1,426 @@
+			onMenuChange(false);
+		}
+	}
+
+	private _renderMenu(results: any[]) {
+		if (!this._open || results.length === 0) {
+			return null;
+		}
+
+		return w('result-menu', <ResultMenuProperties> {
+			bind: this,
+			registry: this.registry,
+			results: results,
+			selectedIndex: this._selectedIndex,
+			getResultLabel: this._getResultLabel,
+			skipResult: this._skipResult,
+			onResultMouseEnter: this._onResultMouseEnter,
+			onResultMouseDown: this._onResultMouseDown,
+			onResultMouseUp: this._onResultMouseUp
+		});
+	}
+
+	private _restoreFocus() {
+		const func = this._focused ? 'focus' : 'blur';
+		this._inputElement[func]();
+	}
+
+	private _selectResult(result: any) {
+		const {
+			autoBlur,
+			onChange
+		} = this.properties;
+
+		this.updateSelectedIndex(Operation.reset);
+		this._open = false;
+		this._focused = !autoBlur;
+		this._ignoreFocus = true;
+		onChange && onChange(this._getResultLabel(result));
+	}
+
+	private _skipResult() {
+		this.updateSelectedIndex(this._direction);
+		this.invalidate();
+	}
+
+	@onPropertiesChanged
+	protected onPropertiesChanged(evt: PropertiesChangeEvent<this, ComboBoxProperties>) {
+		if (!this.registry || includes(evt.changedPropertyKeys, 'customResultItem') || includes(evt.changedPropertyKeys, 'customResultMenu')) {
+			this.registry = this._createRegistry();
+		}
+	}
+
+	protected scrollIntoView(element: HTMLElement) {
+		const menu = <HTMLElement> element.parentElement;
+		// Scroll menu up so top of highlighted result aligns with top of menu container
+		if (element.offsetTop - menu.scrollTop < 0) {
+			menu.scrollTop = element.offsetTop;
+		}
+		// Scroll menu down so bottom of highlighted result aligns with bottom of menu container
+		else if ((element.offsetTop - menu.scrollTop + element.offsetHeight) > menu.clientHeight) {
+			menu.scrollTop = element.offsetTop - menu.clientHeight + element.offsetHeight;
+		}
+	}
+
+	protected updateSelectedIndex(operation?: Operation, index?: number) {
+		const { results } = this.properties;
+
+		if (!results || results.length === 0 || !this._open) {
+			return;
+		}
+
+		if (index !== undefined) {
+			this._selectedIndex = index;
+		}
+		else if (operation !== undefined) {
+			if (operation === Operation.increase || operation === Operation.decrease) {
+				const total = results.length;
+				const base = operation === Operation.increase ? 0 : -1;
+				const current = this._selectedIndex !== undefined ? this._selectedIndex + operation : base;
+				this._selectedIndex = (current + total) % total;
+			}
+			else {
+				this._selectedIndex = undefined;
+			}
+			this._direction = operation;
+		}
+	}
+
+	render(): DNode {
+		const {
+			clearable,
+			disabled,
+			required,
+			invalid,
+			formId,
+			label,
+			readOnly,
+			results = [],
+			inputProperties = {},
+			value = ''
+		} = this.properties;
+
+		const menu = this._renderMenu(results);
+
+		this._notifyMenuChange();
+
+		this._wasOpen = this._open;
+
+		const container = v('div', {
```

Comment:
This element needs a role of `combobox`, `aria-haspopup="true"` (or "listbox"; the docs are a little confusing, and I'd like to test both), and `aria-expanded` set to true/false.

`aria-readonly` and `aria-required` should probably also be set on here, in addition to on the text input.